pages/en/lb3/Discovering-models-from-relational-databases.md

```diff
@@ -23,69 +23,182 @@ This process is called _discovery_ and is supported by the following connectors:
+
+{% include note.html content="
+The code examples in this section are using features introduced in
+Node.js 8.x LTS. Users running on an older version of Node.js need
+to rework the code to use promises (or callbacks) instead of async/await
+for flow control.
+" %}
+
+First of all, we need to import APIs we will be using later:
+
+{% include code-caption.html content="/discover-models.js" %}
+```js
+'use strict';
+
+const loopback = require('loopback');
+const promisify = require('util').promisify;
+const fs = require('fs');
+const writeFile = promisify(fs.writeFile);
+const readFile = promisify(fs.readFile);
+const mkdirp = promisify(require('mkdirp'));
+```
+
+Secondly, we need to initialize the data source backed by the database
+we want to discover our models from. While it's possible to load the full
+application and let loopback-boot load all artifacts (including data sources),
+we recommend a simpler and faster approach of creating the data source
+instance explicitly:
+
+```js
+const DATASOURCE_NAME = 'db';
+const dataSourceConfig = require('./server/datasources.json');
+const db = new loopback.DataSource(dataSourceConfig[DATASOURCE_NAME]);
 ```
 
+Node.js does not allow the main module function to be async,
+the top-level code cannot use `await` keyword for flow control.
+As a workaround, we need to wrap the discovery code in a helper function
+that's executed using Promise API.
+
+```js
+discover().then(
+  success => process.exit(),
+  error => { console.error('UNHANDLED ERROR:\n', error); process.exit(1); },
+);
+```
+
+With the preparations done, it's time to roll up our sleeves
+and do the actual model discovery.
+
+```js
+async function discover() {
+  // It's important to pass the same "options" object to all calls
+  // of dataSource.discoverSchemas(), it allows the method to cache
+  // discovered related models
+  const options = {relations: true};
+
+  // Discover models and relations
+  const salariesSchemas = await db.discoverSchemas('salaries', options);
+  const employeesSchemas = await db.discoverSchemas('employees', options);
+
+  // Create model definition files
+  await mkdirp('common/models');
+  await writeFile(
+    'common/models/salaries.json',
+    JSON.stringify(salariesSchemas['employees.salaries'], null, 2)
+  );
+  await writeFile(
+    'common/models/employees.json',
+    JSON.stringify(salariesSchemas['employees.employees'], null, 2)
+  );
+
+  // Expose models via REST API
+  const configJson = await readFile('server/model-config.json', 'utf-8');
+  console.log('MODEL CONFIG', configJson);
+  const config = JSON.parse(configJson);
+  config.Salaries = {dataSource: DATASOURCE_NAME, public: true};
+  config.Employees = {dataSource: DATASOURCE_NAME, public: true};
+  await writeFile(
+    'server/model-config.json',
+    JSON.stringify(config, null, 2)
+  );
+}
+```
+
+Next steps:
+
+ 1. Run the discovery script to add the discovered models to your project
```

Comment:
I'm really not sure what the context to this step here is. We've managed to create a config file from the schemas saved in the database, but what purpose does that serve? What have we done so far that allows us to `discover and save model definitions`? What is a `discovery script`?

Reply:
Good catch!

In the first part of this section, I have described a discovery script - how to write the code.

In this step, I am telling our users to run the script they have written. (Compare this to the previous section, where the boot script is automatically run by `node .`.)

How can I make this intent more clear? Should I add a "step 0" saying something like "save the source code shown above to a file in your project's root directory"?

> We've managed to create a config file from the schemas saved in the database, but what purpose does that serve? 

We created model config files in such way that application is going to automatically pick these models during app boot (via loopback-boot). 

> What have we done so far that allows us to discover and save model definitions? Is the discovery script the code snippet from the previous section?

So far, the documentation has described how to write a script for discovering models and updating app config to add these models and expose them via app's REST API.

The discovery script is a concatenation of all code snippets from this section.

@shimks could you please help me with coming up with a better way how to express this information in the doc page? Feel free to make the edits directly in the file and push them as a new commit to this feature branch.

Reply:
> Should I add a "step 0" saying something like "save the source code shown above to a file in your project's root directory"?

Reply:
> Should I add a "step 0" saying something like "save the source code shown above to a file in your project's root directory"?

Sounds reasonable to me (that's what I have done).

Reply:
I think it's sufficient enough to add in some words to the first step:
`Run the discovery script we've just written to add the discovered models ...`. In hindsight though, it feels like what the discover script was is really obvious now, so feel free to just leave it as it is.